### DISCUSSION TITLE
Question: How can we setup AD authentication?

### DISCUSSION THREAD

**Discussion by Bahu2016** [fastapi/fastapi#8419] [Questions]
### First Check

- [ ] I added a very descriptive title to this issue.
- [ ] I used the GitHub search to find a similar issue and didn't find it.
- [x] I searched the FastAPI documentation, with the integrated search.
- [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
- [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [ ] I commit to help with one of those options 👆

### Example Code

```python
na
```

### Description

Technical document to setup AD authentication or sso

### Operating System

Linux

### Operating System Details

Centos 7.X

### FastAPI Version

0.87.0

### Python Version

3.10.8

### Additional Context

Looking technical document how to enable AD authentication.

**Comment by bitplane** [comment]
Are you hosting in Azure? If so does this help?

https://intility.github.io/fastapi-azure-auth/

**Comment by extreme4all** [comment]
@Bahu2016 you could probably use an external library like https://pypi.org/project/ldap3/
To use the `ldap3 `library to set up AD integration in FastAPI, you can use the `Connection `and `Server `classes to establish a connection to the AD server and authenticate users.
In this example, the `Server `class is used to create a server instance for the AD server, and the `Connection `class is used to establish a connection and authenticate the user. The user and password parameters are used to specify the username and password for the user to authenticate. If the authentication is successful, the bind method will return True, otherwise it will return False.

After a user successfully logs in to your FastAPI application, you may want to keep track of their session to store information about the user, such as their username, permissions, and preferences. This can be useful for implementing features such as user authentication and authorization, and for maintaining user-specific state in your application.

To keep a session in FastAPI, you can use the SessionMiddleware provided by the fastapi module. This middleware uses cookies to store session data, and it provides methods for accessing and modifying the session data.

Here is an example of how to use ldap3 & SessionMiddleware in FastAPI: (untested, because i don't have an AD)

```py
from fastapi import FastAPI, SessionMiddleware
from ldap3 import Connection, Server

app = FastAPI()
app.add_middleware(SessionMiddleware, secret_key="ChangeY0urS3creT!")

@app.on_event("startup")
async def startup():
    global ad_server
    ad_server = Server("ad.example.com", port=389, use_ssl=False)

@app.post("/login")
async def login(request, username: str, password: str):
    conn = Connection(ad_server, user=username, password=password)
    if conn.bind():
        # Store the user's username in the session
        request.session["username"] = username
        # Store the user's name and email in the session
        user = conn.search(f"cn={username}", attributes=["givenName", "mail"])
        request.session["name"] = user["givenName"]
        request.session["email"] = user["mail"]
        # Store the user's groups in the session
        groups = conn.search(f"cn={username}", attributes=["memberOf"])
        request.session["groups"] = groups["memberOf"]
        return {"message": "Successfully authenticated"}
    else:
        return {"message": "Invalid username or password"}

@app.get("/user")
async def get_user(request):
    username = request.session.get("username")
    name = request.session.get("name")
    email = request.session.get("email")
    groups = request.session.get("groups")
    return {
        "username": username,
        "name": name,
        "email": email,
        "groups": groups,
    }
```
In this example, the `SessionMiddleware `is added to the FastAPI application using the `add_middleware `method. The `secret_key `parameter is used to specify a secret key that will be used to encrypt the session data.

In the `login `route, the `request `object is used to access the session data, and the `username `is stored in the session using the` request.sessio`n dictionary.

In the `get_user `route, the `username `is retrieved from the session using the `request.session.get(`) method, and it is returned in the response.

**Comment by davidmathew** [comment]
Can we configure the OAuth2AuthorizationCodeBearer class with the required authorizationUrl and TokenUrl?

oauth2_scheme = OAuth2AuthorizationCodeBearer(
    authorizationUrl="https://login.microsoftonline.com/Tenant-ID/oauth2/v2.0/authorize", 
    tokenUrl="https://login.microsoftonline.com/Tenant-ID/oauth2/v2.0/token",scopes={"openid":"openid"})

I did try this with the redirect URI (http://localhost:8181/docs/oauth2-redirect) setup with Web  Platform Authentication in the App registration. But I was the getting the error as 

Auth ErrorError: response status is 400, error: invalid_request, description: AADSTS9002326: Cross-origin token redemption is permitted only for the 'Single-Page Application' client-type. Request origin: 'http://localhost:8181'.

Is OAuth2AuthorizationCodeBearer a possibility or do we need to go with lib such as [FastAPI-Azure-Auth](https://intility.github.io/fastapi-azure-auth/)

**Comment by RetributionByRevenue** [comment]
i made a github repo about fastapi auth with windows active directory domain controller 
<pre>
from fastapi import FastAPI, Depends, HTTPException
from fastapi.security import HTTPBasic, HTTPBasicCredentials
from ldap3 import Server, Connection, ALL

def LDAP_AUTH(domain,username,password):
    didConnect=False
    try:
        # Define the server and connection settings
        server = Server(f"ldap://{domain}", get_info=ALL)
        conn = Connection(server, user=f"{username}@{domain}", password=password, auto_bind=True)
        # Attempt to bind (authenticate) the user
        conn.bind()
        # Check if the bind was successful
        if conn.result['result'] == 0:
            print("Authentication successful")
            didConnect = True
    except:
            print("Authentication failed")
    finally:
        # Don't forget to close the connection when you're done
        try:
            conn.unbind()
        except:
            ''
    return didConnect

app = FastAPI()

# Define the HTTPBasic authentication scheme
security = HTTPBasic()

# Dependency to check LDAP authentication
def check_ldap_auth(credentials: HTTPBasicCredentials = Depends(security)):
    domain = "mycompany.com"
    if not LDAP_AUTH(domain, credentials.username, credentials.password):
        raise HTTPException(status_code=401, detail="Invalid credentials")
    return credentials.username

# Example protected route using the dependency
@app.get("/protected")
async def protected_route(username: str = Depends(check_ldap_auth)):
    return {"message": f"Hello, {username}! You have access to this protected route."}

</pre>

**Reply by Amiga501** [reply]
Good work - both yourself for this and of course tiangolo with the centrepiece of the whole thing.

I seem to have stumbled on an issue within HTTPBasic though, doesn't like use of €. 
Or it could be be screwing up, but it'll take $ OK.

**Reply by RetributionByRevenue** [reply]
look at this example @Amiga501 
https://github.com/RetributionByRevenue/fastapi-sessions-domaincontroler
